src/crafting.cpp

```diff
@@ -24,6 +24,28 @@
 #include <math.h>    //sqrt
 #include <algorithm> //std::min
 
+enum usage {
+	use_from_map = 1,
+	use_from_player = 2,
+	use_from_both = 1 | 2,
+	use_from_none = 4,
+	cancel = 5
+};
+
+struct comp_selection {
+	usage use_from;
+	item_comp comp;
+
+	comp_selection() : use_from(usage::use_from_none), comp(item_comp("", 0)) { }
+};
+
+struct tool_selection {
+	usage use_from;
+	tool_comp comp;
+
+	tool_selection() : use_from(usage::use_from_none), comp(tool_comp("", 0)) { }
```

Comment:
The constructor could be removed and the members could be initialized in their declaration instead (if needed at all):

```C++
struct foo {
    usage use_from = use_from_none; // Btw. the usage:: prefix is not needed.
    tool_comp comp; // tool_comp actually already has a decent default constructor.
};
```